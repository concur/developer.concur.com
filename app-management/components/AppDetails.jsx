/* eslint-env browser */

import React, { PropTypes } from 'react';
import { Link } from 'react-router';

class AppDetails extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      app: {},
      loading: false,
      error: false,
    };
  }

  componentWillMount() {
    window.fetch(`${process.env.API_SERVER}/apps/${this.props.params.id}`)
      .then(response => response.json())
      .then(app => this.setState({
        app,
        loading: false,
      }))
      .catch(() => this.setState({
        error: true,
        loading: false,
      }));
  }

  render () {
    if (this.state.loading) {
      return (
        <div className="row">
          <div className="col-md-12">
            <p>Loading...</p>
          </div>
        </div>
      );
    }

    if (this.state.error) {
      return (
        <div className="row">
          <div className="col-md-12">
            <p>An error occurred when loading this app</p>
            <Link to="/">Back to My Apps</Link>
          </div>
        </div>
      );
    }

    const { app } = this.state;

    return (
      <div className="row">
        <div className="col-md-12">
          <h2>{app.name}</h2>
          <Link to="/">&larr; Back to My Apps</Link>
          <div className="row">
            <p>{app.description}</p>
          </div>
        </div>
      </div>
    );
  }
}

AppDetails.propTypes = {
  params: PropTypes.shape({
    id: PropTypes.string.isRequired,
  }),
};

export default AppDetails;
