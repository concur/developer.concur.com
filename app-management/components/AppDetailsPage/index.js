/* eslint-env browser */

import { connect } from 'react-redux';
import AppDetailsPage from './AppDetailsPage';
import { fetchAppDetails, updateAppDetails, showSecret } from '../../actions/appDetails';

export function mapStateToProps(state) {
  return state.appDetails;
}

export function mapDispatchToProps(dispatch) {
  return {
    fetchAppDetails: id => dispatch(fetchAppDetails(id)),
    handleSubmit: app => dispatch(updateAppDetails(app)),
    generateSecret: (id) => {
      const confirmation = window.confirm('Are you sure? This will clear your previous app secret.');
      if (confirmation) {
        dispatch(showSecret());
        console.log(`TBD: Generate secret for app ${id}`);
      }
    },
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AppDetailsPage);
