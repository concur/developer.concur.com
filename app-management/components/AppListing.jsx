import React from 'react';
import { Link } from 'react-router';

class AppListing extends React.Component {
  constructor(props) {
    super(props);
    const apps = [
      {
        id: '2342asdfas',
        name: 'Starbucks',
      },
      {
        id: '234asdfasdf',
        name: 'Uber',
      },
      {
        id: 'asldfkajsdflj',
        name: 'My App',
      },
    ];
    this.state = { apps };
  }

  render () {
    const appItems = this.state.apps.map((app) => {
      const { id, name } = app;
      return (
        <div className="well" key={id}>
          <h2>{name}</h2>
          <Link to="" className="btn-u btn-u-green">Details</Link>
          <Link to="" className="btn-u btn-u-blue">Edit</Link>
        </div>
      );
    });

    return (
      <div className="row">
        <div className="col-md-12">
          <h1>My Apps</h1>
          <div className="row">
            {appItems}
          </div>
        </div>
      </div>
    );
  }
}

export default AppListing;
