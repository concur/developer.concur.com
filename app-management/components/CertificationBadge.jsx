import React, { PropTypes } from 'react';

const CertificationBadge = ({ certified }) => (
  certified ? (
    <small className="certification-badge certified">
      <i className="fa fa-check-circle" aria-hidden="true" /> CERTIFIED
    </small>
  ) : (
    <small className="certification-badge not-certified">
      <i className="fa fa-times-circle" aria-hidden="true" /> NOT CERTIFIED
    </small>
  )
);

CertificationBadge.propTypes = {
  certified: PropTypes.bool,
};

export default CertificationBadge;
