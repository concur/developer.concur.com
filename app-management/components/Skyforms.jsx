/*
 * Wrapper components to render SkyForm fields with validation errors
 */

import React, { PropTypes } from 'react';
import Select from 'react-select';
import { Field } from 'redux-form';

// Wrap each render component with a label and Skyform classes needed for errors
export const FieldWrapper = ({ name, label, touched, error, classNames, children }) => (
  <div>
    <label className="label" htmlFor={name}>{label}</label>
    <div className={touched && error ? `${classNames} state-error` : classNames}>
      {children}
    </div>
    {touched && error && <em htmlFor={name} className="invalid">{error}</em>}
  </div>
);

FieldWrapper.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  touched: PropTypes.bool.isRequired,
  error: PropTypes.any,
  classNames: PropTypes.string,
  children: PropTypes.any.isRequired,
};

export const renderInput = ({ input, name, type, label, meta: { touched, error }, children }) => (
  <FieldWrapper name={name} label={label} touched={touched} error={error} classNames="input">
    <input {...input} id={name} type={type} />
    {children}
  </FieldWrapper>
);

renderInput.propTypes = {
  input: PropTypes.object.isRequired,
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  meta: PropTypes.object.isRequired,
  children: PropTypes.any,
};

export const renderTextarea = ({ input, name, label, meta: { touched, error } }) => (
  <FieldWrapper name={name} label={label} touched={touched} error={error} classNames="textarea">
    <textarea {...input} id={name} />
  </FieldWrapper>
);

renderTextarea.propTypes = {
  input: PropTypes.object.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  meta: PropTypes.object.isRequired,
};

export const renderSelect = ({ input, name, label, options, meta: { touched, error } }) => (
  <FieldWrapper name={name} label={label} touched={touched} error={error}>
    <Select
      {...input}
      id={name}
      options={options}
      onBlur={() => input.onBlur(input.value)}
    />
  </FieldWrapper>
);

renderSelect.propTypes = {
  input: PropTypes.object.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  options: PropTypes.array.isRequired,
  meta: PropTypes.object.isRequired,
};

export const renderMultiSelect = ({ input, name, label, options, meta: { touched, error } }) => (
  <FieldWrapper name={name} label={label} touched={touched} error={error}>
    <Select
      {...input}
      id={name}
      multi
      options={options}
      onBlur={() => input.onBlur(input.value)}
    />
  </FieldWrapper>
);

renderMultiSelect.propTypes = {
  input: PropTypes.object.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  options: PropTypes.array.isRequired,
  meta: PropTypes.object.isRequired,
};

// Due to the checkbox being wrapped by a label, this component doesn't use FieldWrapper
export const renderCheckbox = ({ input, name, type, meta: { touched, error }, children }) => (
  <div>
    <label className={touched && error ? 'checkbox state-error' : 'checkbox'} htmlFor={name}>
      <input {...input} id={name} type={type} />
      <i />
      {children}
    </label>
    {touched && error && <em htmlFor={name} className="invalid">{error}</em>}
  </div>
);

renderCheckbox.propTypes = {
  input: PropTypes.object.isRequired,
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  meta: PropTypes.object.isRequired,
  children: PropTypes.any,
};

export const renderUris = ({ fields }) => (
  <div>
    {fields.map((uri, idx) =>
      <div key={idx}>
        <Field
          component={renderInput}
          type="url"
          name={uri}
          label={`Redirect URI ${idx + 1}`}
          placeholder="Redirect URI"
        >
          {idx ? (
            <i
              className="icon-append fa fa-minus-circle"
              onClick={() => fields.remove(idx)}
            />
          ) : null}
        </Field>
      </div>
    )}
    <div className="add-uri">
      <button type="button" className="button" onClick={() => fields.push('')}>Add URI</button>
    </div>
  </div>
);

renderUris.propTypes = {
  fields: PropTypes.object,
};
