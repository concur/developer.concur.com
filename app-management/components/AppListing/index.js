import { connect } from 'react-redux';
import AppListing from './AppListing';
import { fetchAppListing } from '../../actions/appListing';

export function mapStateToProps(state) {
  return state.appListing;
}

export function mapDispatchToProps(dispatch) {
  return {
    fetchAppListing: () => dispatch(fetchAppListing()),
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AppListing);
