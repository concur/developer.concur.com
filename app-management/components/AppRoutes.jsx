import React, { PropTypes } from 'react';
import { Router, Route, IndexRoute, hashHistory } from 'react-router';

import PageContainer from './PageContainer';
import AppListing from './AppListing';
import AppDetailsPage from './AppDetailsPage';
import NewAppPage from './NewAppPage';
import LoginPage from './LoginPage';
import SignupPage from './SignupPage';
import LogoutPage from './LogoutPage';
import NotFound from './NotFound';

import { logout } from '../actions/auth';

class AppRoutes extends React.Component {
  constructor(props) {
    super(props);

    this.isAuthenticated = this.isAuthenticated.bind(this);
    this.requireAuth = this.requireAuth.bind(this);
    this.showAuthForms = this.showAuthForms.bind(this);
    this.logout = this.logout.bind(this);
  }

  isAuthenticated() {
    const { store } = this.props;
    return !!store.getState().auth.token;
  }

  requireAuth(nextState, replaceState) {
    if (!this.isAuthenticated()) {
      replaceState('/login');
    }
  }

  showAuthForms(nextState, replaceState) {
    if (this.isAuthenticated()) {
      replaceState('/');
    }
  }

  logout(nextState, replaceState) {
    if (this.isAuthenticated()) {
      this.props.store.dispatch(logout());
    } else {
      replaceState('/login');
    }
  }

  render () {
    return (
      <Router history={hashHistory}>
        <Route path="/" component={PageContainer}>
          <IndexRoute component={AppListing} onEnter={this.requireAuth} />
          <Route path="new" component={NewAppPage} onEnter={this.requireAuth} />
          <Route path="details/:id" component={AppDetailsPage} onEnter={this.requireAuth} />
          <Route path="login" component={LoginPage} onEnter={this.showAuthForms} />
          <Route path="signup" component={SignupPage} onEnter={this.showAuthForms} />
          <Route path="logout" component={LogoutPage} onEnter={this.logout} />
          <Route path="*" component={NotFound} />
        </Route>
      </Router>
    );
  }
}

AppRoutes.propTypes = {
  store: PropTypes.shape({
    dispatch: PropTypes.func.isRequired,
  }),
};

export default AppRoutes;
