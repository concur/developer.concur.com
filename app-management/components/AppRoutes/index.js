import AppRoutes from './AppRoutes';

export default AppRoutes;
