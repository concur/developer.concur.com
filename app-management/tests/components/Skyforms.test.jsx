import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { FieldWrapper, renderSelect, renderCheckbox, renderUris } from '../../components/Skyforms';

describe('<FieldWrapper />', () => {
  const props = {
    name: 'My field',
    label: 'my-field',
    touched: true,
    error: ['Required'],
    classNames: 'my-field',
  };
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <FieldWrapper {...props}>
        <input type="text"/>
      </FieldWrapper>
    );
  });

  it('should display the state-error class when there is an error', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
