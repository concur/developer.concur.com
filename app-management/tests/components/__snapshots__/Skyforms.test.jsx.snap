exports[`<FieldWrapper /> should display the state-error class when there is an error 1`] = `
<div>
  <label
    className="label"
    htmlFor="My field">
    my-field
  </label>
  <div
    className="my-field state-error">
    <input
      type="text" />
  </div>
  <em
    className="invalid"
    htmlFor="My field">
    Required
  </em>
</div>
`;
